src/main.rs

```diff
@@ -201,6 +202,125 @@ async fn make_rpc_client(
     Ok(client)
 }
 
+struct ProgressBarState {
+    mp: MultiProgress,
+    pbs: HashMap<u64, ProgressBar>,
+}
+
+impl ProgressBarState {
+    fn new() -> Self {
+        Self {
+            mp: MultiProgress::new(),
+            pbs: HashMap::new(),
+        }
+    }
+
+    fn found(&mut self, name: String, id: u64, size: u64) {
+        let pb = self.mp.add(ProgressBar::new(size));
+        pb.set_style(ProgressStyle::default_bar()
+            .template("{spinner:.green} [{bar:40.cyan/blue}] {msg} {bytes}/{total_bytes} ({bytes_per_sec}, eta {eta})").unwrap()
+            .progress_chars("=>-"));
+        pb.set_message(name);
+        pb.set_position(0);
+        self.pbs.insert(id, pb);
+    }
+
+    fn progress(&mut self, id: u64, progress: u64) {
+        if let Some(pb) = self.pbs.get_mut(&id) {
+            pb.set_position(progress);
+        }
+    }
+
+    fn done(&mut self, id: u64, _hash: Hash) {
+        if let Some(pb) = self.pbs.remove(&id) {
+            pb.finish_and_clear();
+            self.mp.remove(&pb);
+        }
+    }
+
+    fn all_done(self) {
+        self.mp.clear().ok();
+    }
+
+    fn error(self) {
+        self.mp.clear().ok();
+    }
+}
+
+#[derive(Debug)]
+struct ProvideResponseEntry {
+    pub name: String,
+    pub size: u64,
+}
+
+async fn aggregate_add_response(
+    stream: impl Stream<
+            Item = std::result::Result<
+                ProvideProgress,
+                impl std::error::Error + Send + Sync + 'static,
+            >,
+        > + Unpin,
+) -> anyhow::Result<(Hash, Vec<ProvideResponseEntry>)> {
+    let mut stream = stream;
```

Comment:
you can't declare this mut in the signature?

Reply:
I can, but I don't want to. I find that the mut is entirely an implementation detail and does not belong in the signature.

Reply:
putting `mut` in the "signature" has no effect on the caller at all.  It's a weirdness of Rust syntax where putting `mut` there or not has no effect on the signature at all, it is just something local to the function body.

So you're saying you so strongly disagree with Rust's syntax choices that you prefer to do manual work arounds :smile:  While I agree it's a bit odd, I don't tend to agree with fighting stuff like that.  Once you know how this works I think it's perfectly fine.